proto/api/v1/cloud_source_service.proto

```diff
@@ -0,0 +1,155 @@
+import "gogoproto/gogo.proto";
+
+package v1;
+
+// CloudSource is an integration which provides a source for discovered
+// clusters.
+message CloudSource {
+    string id                            = 1;
+    string name                          = 2;
+    string type                          = 3;
+    bool   skip_test_integration         = 4;
+
+    oneof Config {
+        PaladinCloudConfig paladin_cloud = 5;
+        OCMConfig          ocm           = 6;
+    }
+}
+
+// PaladinCloudConfig provides information required to fetch discovered
+// clusters from Paladin Cloud.
+message PaladinCloudConfig {
+    string endpoint = 1 [(gogoproto.moretags) = 'scrub:"dependent" validate:"nolocalendpoint"'];
+    string token    = 2 [(gogoproto.moretags) = 'scrub:"always"'];
+}
+
+// OCMConfig provides information required to fetch discovered clusters from
+// the OpenShift cluster manager.
+message OCMConfig {
+    string endpoint = 1 [(gogoproto.moretags) = 'scrub:"dependent" validate:"nolocalendpoint"'];
+    string token    = 2 [(gogoproto.moretags) = 'scrub:"always"'];
+}
+
+message CloudSourcesFilter {
+    // Matches cloud sources based on their type.
+    repeated string types = 1;
+}
+
+message CountCloudSourcesRequest {
+    // Filters cloud sources based on the requested fields.
+    CloudSourcesFilter filter = 1;
+}
+
+message CountCloudSourcesResponse {
+    int32 count = 1;
+}
+
+message GetCloudSourceRequest {
+    string id = 1;
+}
+
+message GetCloudSourceResponse {
+    CloudSource item = 1;
+}
+
+message ListCloudSourcesRequest {
+    // Divides the response into chunks.
+    Pagination pagination = 1;
+
+    // Filters cloud sources based on the requested fields.
+    CloudSourcesFilter filter = 2;
+}
+
+message ListCloudSourcesResponse {
+    repeated CloudSource items = 1;
+}
+
+message PostCloudSourceRequest {
+    CloudSource item = 1;
+}
+
+message PostCloudSourceResponse {
+    CloudSource item = 1;
+}
+
+message PutCloudSourceRequest {
+    CloudSource item        = 1;
+    // If true, item must include valid credentials.
+    // If false, the resource must already exist and
+    // credentials in item are ignored.
+    bool update_credentials = 2;
+}
+
+message DeleteCloudSourceRequest {
+    string id = 1;
+}
+
+message TestCloudSourceRequest {
+    CloudSource item        = 1;
+    // If true, item must include valid credentials.
+    // If false, the resource must already exist and
+    // credentials in item are ignored.
+    bool update_credentials = 2;
+}
+
+service CloudSourcesService {
+    // CountCloudSources returns the number of cloud sources after filtering by requested fields.
+    rpc CountCloudSources(CountCloudSourcesRequest) returns (CountCloudSourcesResponse) {
+        option (google.api.http) = {
+            get: "/v1/count/cloudsources"
+        };
+    }
+
+    // GetCloudSource retrieves a cloud source by ID.
+    rpc GetCloudSource(GetCloudSourceRequest) returns (GetCloudSourceResponse) {
```

Comment:
There's `ResourceByID` for that, let's use that.

Reply:
Our new API guidelines recommend against `ResourceByID`, see https://github.com/stackrox/stackrox/blob/master/.github/api_guidelines.md#grpc-message-name.

Reply:
1. Informing, not complaining:
    * Because API token and cluster init bundle are already special cases, I will find everywhere I need to add corresponding special cases for cloud source integrations.
    * IF cloud source were the first exception from pattern of backup integrations, image integrations, notifier integrations, then update to new API guidelines might have been a risk to the very short timeline, because frontend code had been (too) strongly factored to that pattern.
2. Stephan, for anything like this or merge of PATCH into PUT that differs from previous integrations, can you include as comments to this PR an example curl command and response data so I can make sure frontend payloads, responses, and service functions are correct?
3. Although I am personally neutral about `/v1/cloudsource` versus `/v1/cloud-source` if various other updates, like this, then update to naming convention makes sense. For example, Van prefers hyphen.